Sources_v3/StreamChat/Workers/Background/AttachmentUploader.swift

```diff
@@ -0,0 +1,151 @@
+        database.write { [weak self] session in
+            guard
+                let attachmentID = self?.pendingAttachmentIDs.first
+            else { return }
+
+            guard
+                let attachment: _ChatMessageAttachment<ExtraData> = session.attachment(id: attachmentID)?.asModel(),
+                attachment.localState == .pendingUpload,
+                let localFileURL = attachment.localURL
+            else {
+                self?.removeAttachmentIDAndContinue(attachmentID)
+                return
+            }
+
+            guard let fileData = try? Data(contentsOf: localFileURL) else {
+                self?.markAttachment(attachmentID, as: .uploadingFailed) {
+                    self?.removeAttachmentIDAndContinue(attachmentID)
+                }
+                return
+            }
+
+            let fileType = AttachmentFileType(ext: localFileURL.pathExtension)
+
+            let multipartFormData = MultipartFormData(
+                fileData,
+                fileName: attachment.title,
+                mimeType: fileType.mimeType
+            )
+
+            self?.apiClient.uploadFile(
+                endpoint: .uploadAttachment(with: attachment.id, type: attachment.type),
+                multipartFormData: multipartFormData,
+                progress: {
+                    self?.markAttachment(attachmentID, as: .uploading(progress: $0), completion: {})
+                },
+                completion: { result in
+                    self?.database.write({
+                        guard
+                            let attachmentDTO = $0.attachment(id: attachmentID),
+                            let messageDTO = $0.message(id: attachmentID.messageId)
+                        else { return }
+
+                        switch result {
+                        case let .success(payload):
+                            if attachmentDTO.type == AttachmentType.image.rawValue {
+                                attachmentDTO.imageURL = payload.file
+                            } else {
+                                let file = AttachmentFile(
+                                    type: fileType,
+                                    size: Int64(fileData.count),
+                                    mimeType: fileType.mimeType
+                                )
+                                attachmentDTO.file = try JSONEncoder().encode(file)
+                                attachmentDTO.url = payload.file
+                            }
+                            attachmentDTO.localState = .uploaded
+
+                            // Re-assign state to trigger observing FRC
+                            let messageState = messageDTO.localMessageState
+                            messageDTO.localMessageState = messageState
+                        case .failure:
+                            attachmentDTO.localState = .uploadingFailed
+                            messageDTO.localMessageState = .sendingFailed
```

Comment:
This is an interesting question. I'm not sure we should mark the message as failed. That state is reserved for the situation when the actual message sending fails. I'd say we should keep the message state untouched but I don't feel strongly about this.

Reply:
fixed in ad4172515548d622f4423ec8ecaa1c5578267248